Check modification times of included RST files

yongrenjie opened the pull request:
Fixes #17860. Briefly, to reproduce this bug:

1. Create a file `included.rst` with a plot inside it.

2. Create a file `index.rst` containing the include directive
    `.. include:: included.rst`.

3. Build Sphinx documentation.

4. Modify the plotting code inside `included.rst`, but don't touch
   `index.rst`.

5. Rebuild the Sphinx documentation. Because `index.rst` wasn't touched,
   plot_directive thinks that it's up-to-date and thus the image doesn't
   need to be re-created.

This commit fixes the issue by checking the modification times not only
of the source file (`index.rst` in this case), but also of the file(s)
included within it (`included.rst`).

## PR Summary

The commit message contains most of the info, but I attach some extra detail here just in case people come across it and find it useful. There are a couple of ways to get the included files:

**Method 1**

Since docutils 0.17, there is an attribute `state.document.include_log` which is a list of tuples containing the names of any files that were included, which is exactly what we want.

```
[('index.rst', (None, None, None, None), 4.611686018427388e+18),
 ('included.rst', (None, None, None, None), 21)]
```

**Method 2**

Directly inspect `state_machine.input_lines.items`, which is a list of tuples `(source, offset)`. If a RST file is included, then we can get its filename (relative to the top-level Sphinx directory) in `source`. For example, this is what `plot_directive` can when I place `.. include:: included.rst` inside the file `~/test/plot_dir/index.rst`:

```
[('included.rst', 2),
 ('included.rst', 3),
 ('included.rst', 4),
 ('included.rst', 5),
 ('included.rst', 6),
 ('included.rst', 7), 
 ('included.rst', 8),
 ('included.rst', 9),
 ('included.rst', 10),
 ('included.rst', 11),
 ('included.rst', 12),
 ('internal padding after included.rst', 13),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 7),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 8),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 9),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 10),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 11),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 12),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 13),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 14),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 15),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 16),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 17),
 ('/Users/yongrenjie/test/plot_dir/index.rst', 18)
]
```

The issue with this is that there's the 'internal padding after...' line which we would have to filter out, and I'm not sure what other directives (if any) might also add some junk to this list. I got around this by first removing duplicates and then checking which of the entries were valid files.

I feel like Method 2 is a bit fragile, so I made it as a fallback:

```
try:
    # method 1...
except AttributeError:    # indicates docutils <=0.16
    # method 2...
```

I've tested this all the way back to docutils 0.14 and it seems to be pretty reliable.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

- [x] **N/A** Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [x] **N/A** New features are documented, with examples if plot related.
- [x] **N/A** Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [x] **N/A** New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [x] **N/A** API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

tacaswell commented:
Thank you for your work on this!  The approach looks sound to me, however I have a slight concern about the implementation.

How "public" are these functions?  That is, do we expect anyone outside of the internal code to the plot extension to be using them?  These changes introduce backwards-incompatible changes (because it adds positional arguement someplace in the middle).  It would be safer from a back-compat stand point (but less great from a forward looking API point of view) to do the signatures as

```python
def out_of_date(original, derived, includes=None):
    if includes is None:
        includes = []
    ...
```

```python
def render_figures(code, code_path
￼                  output_dir, output_base, context,
￼                  function_name, config, context_reset=False,		￼                   
                   close_figs=False, code_includes=None):
    if code_includes is None:
        code_includes = []
    ...
```
so if anyone _is_ using these we won't break them!

yongrenjie commented:
@tacaswell I didn't think about that possibility, to be honest. I assumed that the implementation of plot_directive was internal and that the user-facing part would be the Sphinx extension itself. However, I entirely agree with the backwards compatibility point, those functions are exposed and can be imported, so there might be someone out there using it. I'll modify it as per your suggestion.

Thanks for taking a look & commenting! 😄

yongrenjie commented:
@tacaswell I updated and rebased, let me know if there's anything else I could change.

QuLogic commented:
I'm not sure if it does check this for non-included files anyway, but can you update the test for this?

yongrenjie commented:
@QuLogic As far as I could tell, `tests/test_sphinxext.py` only runs a single pass of `python -m sphinx`, so there isn't yet anything that tests plot regeneration based on modification time.

Got a bit of a busy time ahead, so I'm not sure I'll be able to add new tests for this right now, sorry—I wouldn't mind giving it a shot in the near-ish future though!

yongrenjie commented:
Sorry for the delay (real life got in the way); but I think this is ready for review. As a summary: I fixed the original bug (#17860), fixed an extra bug along the way (#20523, there's rather more explanation there), and added tests to make sure both behave correctly. PEP8 tests are all fine.

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -742,6 +754,25 @@ def run(arguments, content, options, state_machine, state, lineno):
+    except AttributeError:
+        # the document.include_log attribute only exists in docutils >=0.17,
+        # before that we need to inspect the state machine
+        possible_sources = [os.path.join(setup.confdir, t[0])
+                            for t in state_machine.input_lines.items]
+        source_file_includes = [f for f in set(possible_sources)
+                                if os.path.isfile(f)]
```
```suggestion
        possible_sources = {os.path.join(setup.confdir, t[0])
                            for t in state_machine.input_lines.items}
        source_file_includes = [f for f in possible_sources
                                if os.path.isfile(f)]
```

QuLogic commented on lib/matplotlib/tests/test_sphinxext.py:
```diff
@@ -5,6 +5,7 @@
 from pathlib import Path
 from subprocess import Popen, PIPE
 import sys
+import shutil
```
```suggestion
import shutil
from subprocess import Popen, PIPE
import sys
```

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -429,14 +429,23 @@ def filenames(self):
-    return (not os.path.exists(derived) or
-            (os.path.exists(original) and
-             os.stat(derived).st_mtime < os.stat(original).st_mtime))
+    if includes is None:
+        includes = []
+    files_to_check = [original, *includes]
+
+    def out_of_date_one(original, derived):
+        return (not os.path.exists(derived) or
+                (os.path.exists(original) and
+                 os.stat(derived).st_mtime < os.stat(original).st_mtime))
+
+    return any(out_of_date_one(f, derived) for f in files_to_check)
```
This can be optimized slightly:
```suggestion
    if not os.path.exists(derived):
        return True

    if includes is None:
        includes = []
    files_to_check = [original, *includes]

    def out_of_date_one(original, derived_mtime):
        return (os.path.exists(original) and
                derived_mtime < os.stat(original).st_mtime)

    derived_mtime = os.stat(derived).st_mtime
    return any(out_of_date_one(f, derived_mtime) for f in files_to_check)
```
Maybe don't even need the function now.

QuLogic commented on lib/matplotlib/tests/test_sphinxext.py:
```diff
@@ -13,27 +14,21 @@
-                 env={**os.environ, "MPLBACKEND": ""})
-    out, err = proc.communicate()
+    source_dir = Path(tmpdir) / 'src'
+    shutil.copytree(str(Path(__file__).parent / 'tinypages'), str(source_dir))
```
```suggestion
    shutil.copytree(Path(__file__).parent / 'tinypages', source_dir)
```

QuLogic commented on lib/matplotlib/tests/test_sphinxext.py:
```diff
@@ -62,3 +58,47 @@ def plot_file(num):
+    assert filecmp.cmp(range_6, plot_file(17))
+
+    # Modify the included plot
+    with open(str(source_dir / 'included_plot_21.rst'), 'r') as file:
+        contents = file.read()
+    contents = contents.replace('plt.plot(range(6))', 'plt.plot(range(4))')
+    with open(str(source_dir / 'included_plot_21.rst'), 'w') as file:
+        file.write(contents)
```
```suggestion
    contents = (source_dir / 'included_plot_21.rst').read_text()
    contents = contents.replace('plt.plot(range(6))', 'plt.plot(range(4))')
    (source_dir / 'included_plot_21.rst').write_text(contents)
```

yongrenjie commented:
Thanks @QuLogic for your corrections!

I think that the `out_of_date` function is still needed for two cases - (1) is when the plot comes from a python file `.. plot:: my_plot.py` and (2) for the included rst files (which is what I tried to fix). Unless I'm missing something, which might well be the case.

QuLogic commented:
I just meant it could be inlined.

QuLogic commented:
Thanks @yongrenjie! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

yongrenjie commented:
Thank you! 😄 BTW I see what you meant now with inlining... I could put in another PR if you like.

```python
def out_of_date(original, derived, includes=None):
    if not os.path.exists(derived):
        return True
    else:
        derived_mtime = os.stat(derived).st_mtime
    if includes is None:
        includes = []
    files_to_check = [original, *includes]
    return any((derived_mtime < os.stat(f).st_mtime) for f in files_to_check)
```

QuLogic commented:
Sure, feel free to do so.
